File: android/app/src/main/java/com/magav/app/service/SmsReminderService.kt
package com.magav.app.service

import com.magav.app.db.MagavDatabase
import com.magav.app.db.entity.SchedulerConfigEntity
import com.magav.app.db.entity.SchedulerRunLogEntity
import com.magav.app.db.entity.LocationEntity
import com.magav.app.db.entity.SmsLogEntity
import com.magav.app.db.entity.VolunteerEntity
import com.magav.app.sms.SmsProvider
import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import com.magav.app.util.ReminderTypes
import com.magav.app.util.SmsStatuses
import com.magav.app.util.toIsoRange

data class SmsSummary(val totalEligible: Int, val smsSent: Int, val smsFailed: Int)

class SmsReminderService(
    private val database: MagavDatabase,
    private val smsProvider: SmsProvider
) {
    suspend fun execute(config: SchedulerConfigEntity, targetDate: LocalDate): SmsSummary {
        val (targetDateStart, targetDateEnd) = targetDate.toIsoRange()
        val targetDateStr = targetDate.format(DateTimeFormatter.ISO_LOCAL_DATE)
        val reminderType = config.reminderType

        android.util.Log.d("SmsReminder", "execute: config=${config.id}, type=$reminderType, target=$targetDateStr, range=[$targetDateStart, $targetDateEnd)")

        // Resolve message template once before the loop
        val messageTemplate = database.messageTemplateDao().getById(config.messageTemplateId)
        if (messageTemplate == null) {
            android.util.Log.e("SmsReminder", "MessageTemplate ${config.messageTemplateId} not found for config ${config.id}")
            return SmsSummary(0, 0, 0)
        }

        val shifts = database.shiftDao().getByDateRange(targetDateStart, targetDateEnd)
        android.util.Log.d("SmsReminder", "Found ${shifts.size} shifts for $targetDateStr")

        // Bulk loads (replaces N+1 per-shift queries)
        val volunteerMap: Map<Int, VolunteerEntity>
        val sentShiftIds: Set<Int>
        val locationMap: Map<Int, LocationEntity>
        if (shifts.isNotEmpty()) {
            volunteerMap = database.volunteerDao().getAll().associateBy { it.id }
            val shiftIds = shifts.map { it.id }
            sentShiftIds = database.smsLogDao().getSuccessfulByShiftIdsAndReminderType(shiftIds, reminderType)
                .map { it.shiftId }.toSet()
            locationMap = if (reminderType == ReminderTypes.SAME_DAY) {
                database.locationDao().getAll().associateBy { it.id }
            } else emptyMap()
        } else {
            volunteerMap = emptyMap()
            sentShiftIds = emptySet()
            locationMap = emptyMap()
        }

        var totalEligible = 0
        var smsSent = 0
        var smsFailed = 0

        for (shift in shifts) {
            val volId = shift.volunteerId ?: continue
            val volunteer = volunteerMap[volId]
            if (volunteer == null) {
                android.util.Log.w("SmsReminder", "Volunteer ${shift.volunteerId} not found for shift ${shift.id}")
                continue
            }

            if (volunteer.approveToReceiveSms != 1) {
                android.util.Log.d("SmsReminder", "Skip ${volunteer.mappingName}: not approved")
                continue
            }
            if (volunteer.mobilePhone.isNullOrBlank()) {
                android.util.Log.d("SmsReminder", "Skip ${volunteer.mappingName}: no phone")
                continue
            }

            if (shift.id in sentShiftIds) {
                android.util.Log.d("SmsReminder", "Skip ${volunteer.mappingName}: already sent")
                continue
            }

            totalEligible++

            try {
                var message = buildMessage(
                    messageTemplate.content, shift.shiftName, shift.carId,
                    volunteer.mappingName, targetDate
                )
                if (reminderType == ReminderTypes.SAME_DAY) {
                    val location = shift.locationId?.let { locationMap[it] }
                    val locName = location?.name ?: shift.customLocationName
                    val locNav = location?.navigation ?: shift.customLocationNavigation
                    val locCity = location?.city
                    message += buildLocationText(locName, locCity, locNav)
                }
                android.util.Log.d("SmsReminder", "Sending SMS #$totalEligible to ${volunteer.mappingName} (${volunteer.mobilePhone})")
                val result = smsProvider.sendSms(volunteer.mobilePhone, message)
                android.util.Log.d("SmsReminder", "SMS result: success=${result.success}, error=${result.error}")

                val now = Instant.now().toString()
                val smsLog = SmsLogEntity(
                    shiftId = shift.id,
                    sentAt = now,
                    status = if (result.success) SmsStatuses.SUCCESS else SmsStatuses.FAIL,
                    error = result.error,
                    reminderType = reminderType
                )
                database.smsLogDao().insert(smsLog)

                if (result.success) {
                    smsSent++
                    // Update SmsSentAt
                    database.shiftDao().update(shift.copy(smsSentAt = now))
                } else {
                    smsFailed++
                }

                // Add delay between SMS to avoid carrier rate limiting
                if (totalEligible > 1) {
                    kotlinx.coroutines.delay(500)
                }
            } catch (e: Exception) {
                smsFailed++
                try {
                    database.smsLogDao().insert(
                        SmsLogEntity(
                            shiftId = shift.id,
                            sentAt = Instant.now().toString(),
                            status = SmsStatuses.FAIL,
                            error = "שגיאה פנימית",
                            reminderType = reminderType
                        )
                    )
                } catch (_: Exception) {
                }
            }
        }

        // Determine status
        val status = when {
            totalEligible == 0 -> "Completed"
            smsFailed == 0 -> "Completed"
            smsSent == 0 -> "Failed"
            else -> "Partial"
        }
        val runError = if (smsFailed > 0) "$smsFailed הודעות נכשלו" else null

        android.util.Log.d("SmsReminder", "Summary: eligible=$totalEligible, sent=$smsSent, failed=$smsFailed, status=$status")

        // Insert SchedulerRunLog
        try {
            database.schedulerRunLogDao().insert(
                SchedulerRunLogEntity(
                    configId = config.id,
                    reminderType = reminderType,
                    ranAt = Instant.now().toString(),
                    targetDate = targetDateStr,
                    totalEligible = totalEligible,
                    smsSent = smsSent,
                    smsFailed = smsFailed,
                    status = status,
                    error = runError
                )
            )
        } catch (_: Exception) {
            // UNIQUE constraint violation = already ran
        }

        return SmsSummary(totalEligible, smsSent, smsFailed)
    }

    companion object {
        fun buildMessage(
            template: String, shiftName: String, carId: String,
            volunteerName: String, targetDate: LocalDate
        ): String {
            val dateStr = "${targetDate.dayOfMonth.toString().padStart(2, '0')}/" +
                "${targetDate.monthValue.toString().padStart(2, '0')}/${targetDate.year}"
            val dayName = getHebrewDayName(targetDate.dayOfWeek)

            return template
                .replace("{שם}", volunteerName)
                .replace("{שם מלא}", volunteerName)
                .replace("{תאריך}", dateStr)
                .replace("{יום}", dayName)
                .replace("{משמרת}", shiftName)
                .replace("{רכב}", carId)
        }

        fun buildLocationText(name: String?, city: String?, navigation: String?): String {
            if (name.isNullOrEmpty()) return ""
            val text = if (!city.isNullOrEmpty()) "\nהניידת נמצאת ב$city ($name)"
                       else "\nהניידת נמצאת אצל $name"
            return if (!navigation.isNullOrEmpty()) "$text ,נווט $navigation" else text
        }

        fun getHebrewDayName(day: DayOfWeek): String = when (day) {
            DayOfWeek.SUNDAY -> "יום א׳"
            DayOfWeek.MONDAY -> "יום ב׳"
            DayOfWeek.TUESDAY -> "יום ג׳"
            DayOfWeek.WEDNESDAY -> "יום ד׳"
            DayOfWeek.THURSDAY -> "יום ה׳"
            DayOfWeek.FRIDAY -> "יום ו׳"
            DayOfWeek.SATURDAY -> "שבת"
        }
    }
}

File: web/server/Magav.Server/Services/Sms/SmsReminderService.cs
using Magav.Common;
using Magav.Common.Models;
using Magav.Server.Database;
using Microsoft.Extensions.Logging;

namespace Magav.Server.Services.Sms;

public class SmsReminderService
{
    private readonly MagavDbManager _db;
    private readonly ISmsProvider _smsProvider;
    private readonly ILogger<SmsReminderService> _logger;

    public SmsReminderService(
        MagavDbManager db,
        ISmsProvider smsProvider,
        ILogger<SmsReminderService> logger)
    {
        _db = db;
        _smsProvider = smsProvider;
        _logger = logger;
    }

    public async Task ExecuteAsync(SchedulerConfig config, DateTime targetDate, CancellationToken ct)
    {
        var targetDateStart = targetDate.Date.ToString("o");
        var targetDateEnd = targetDate.Date.AddDays(1).ToString("o");
        var targetDateStr = targetDate.Date.ToString("yyyy-MM-dd");
        var reminderType = config.ReminderType;

        _logger.LogInformation(
            "Scheduler run starting: ConfigId={ConfigId}, ReminderType={ReminderType}, TargetDate={TargetDate}",
            config.Id, reminderType, targetDateStr);

        // Query eligible shifts: shifts on target date with approved volunteers,
        // excluding those that already have a successful SmsLog for this ReminderType
        var eligibleShifts = await _db.Db.FetchAsync<ShiftVolunteerDto>(
            @"SELECT s.Id AS ShiftId, s.ShiftDate, s.ShiftName, s.CarId,
                     v.Id AS VolunteerId, v.FirstName, v.LastName, v.MappingName, v.MobilePhone,
                     s.LocationId,
                     COALESCE(l.Name, s.CustomLocationName) AS LocationName,
                     COALESCE(l.Navigation, s.CustomLocationNavigation) AS LocationNavigation,
                     l.City AS LocationCity
              FROM Shifts s
              JOIN Volunteers v ON s.VolunteerId = v.Id
              LEFT JOIN Locations l ON s.LocationId = l.Id
              WHERE s.ShiftDate >= @0 AND s.ShiftDate < @1
                AND v.ApproveToReceiveSms = 1
                AND v.MobilePhone IS NOT NULL
                AND v.MobilePhone != ''
                AND NOT EXISTS (
                    SELECT 1 FROM SmsLog sl
                    WHERE sl.ShiftId = s.Id
                      AND sl.ReminderType = @2
                      AND sl.Status = @3
                )",
            targetDateStart, targetDateEnd, reminderType, MagavConstants.SmsStatuses.Success);

        var totalEligible = eligibleShifts.Count;
        var smsSent = 0;
        var smsFailed = 0;
        string? runError = null;

        _logger.LogInformation("Found {Count} eligible shifts for {ReminderType} on {TargetDate}",
            totalEligible, reminderType, targetDateStr);

        // Resolve message template once before the loop
        var template = await _db.MessageTemplates.GetByIdAsync(config.MessageTemplateId);
        if (template == null)
        {
            _logger.LogError("MessageTemplate {Id} not found for config {ConfigId}",
                config.MessageTemplateId, config.Id);
            await _db.SchedulerRunLog.InsertAsync(new SchedulerRunLog
            {
                ConfigId = config.Id,
                ReminderType = config.ReminderType,
                RanAt = DateTime.UtcNow.ToString("o"),
                TargetDate = targetDate.Date.ToString("yyyy-MM-dd"),
                TotalEligible = 0,
                SmsSent = 0,
                SmsFailed = 0,
                Status = "Failed",
                Error = "תבנית הודעה לא נמצאה"
            });
            return;
        }

        foreach (var shift in eligibleShifts)
        {
            if (ct.IsCancellationRequested) break;

            try
            {
                var message = BuildMessage(template.Content, shift, targetDate);
                if (reminderType == MagavConstants.ReminderTypes.SameDay)
                    message += BuildLocationText(shift);
                var result = await _smsProvider.SendSmsAsync(shift.MobilePhone!, message);

                // Write SmsLog entry
                var smsLog = new SmsLog
                {
                    ShiftId = shift.ShiftId,
                    SentAt = DateTime.UtcNow,
                    Status = result.Success ? MagavConstants.SmsStatuses.Success : MagavConstants.SmsStatuses.Fail,
                    Error = result.Error,
                    ReminderType = reminderType
                };
                await _db.SmsLog.InsertAsync(smsLog);

                // Update SmsSentAt on success (general indicator)
                if (result.Success)
                {
                    smsSent++;
                    await _db.Db.ExecuteQueryAsync(
                        "UPDATE Shifts SET SmsSentAt = @0 WHERE Id = @1",
                        DateTime.UtcNow.ToString("o"), shift.ShiftId);
                }
                else
                {
                    smsFailed++;
                    _logger.LogWarning("SMS failed for ShiftId={ShiftId}: {Error}",
                        shift.ShiftId, result.Error);
                }
            }
            catch (Exception ex)
            {
                smsFailed++;
                _logger.LogError(ex, "Error sending SMS for ShiftId={ShiftId}", shift.ShiftId);

                // Still log the failure in SmsLog
                try
                {
                    var smsLog = new SmsLog
                    {
                        ShiftId = shift.ShiftId,
                        SentAt = DateTime.UtcNow,
                        Status = MagavConstants.SmsStatuses.Fail,
                        Error = "שגיאה פנימית",
                        ReminderType = reminderType
                    };
                    await _db.SmsLog.InsertAsync(smsLog);
                }
                catch (Exception logEx)
                {
                    _logger.LogError(logEx, "Failed to write SmsLog for ShiftId={ShiftId}", shift.ShiftId);
                }
            }
        }

        // Determine run status
        var status = totalEligible == 0 ? "Completed"
            : smsFailed == 0 ? "Completed"
            : smsSent == 0 ? "Failed"
            : "Partial";

        if (smsFailed > 0)
            runError = $"{smsFailed} הודעות נכשלו";

        // Insert SchedulerRunLog (UNIQUE constraint prevents duplicates)
        var runLog = new SchedulerRunLog
        {
            ConfigId = config.Id,
            ReminderType = reminderType,
            RanAt = DateTime.UtcNow.ToString("o"),
            TargetDate = targetDateStr,
            TotalEligible = totalEligible,
            SmsSent = smsSent,
            SmsFailed = smsFailed,
            Status = status,
            Error = runError
        };

        var inserted = await _db.SchedulerRunLog.InsertAsync(runLog);
        if (inserted == null)
        {
            _logger.LogWarning(
                "SchedulerRunLog already exists for ConfigId={ConfigId}, TargetDate={TargetDate}, ReminderType={ReminderType}",
                config.Id, targetDateStr, reminderType);
        }

        _logger.LogInformation(
            "Scheduler run completed: ConfigId={ConfigId}, Status={Status}, Sent={Sent}, Failed={Failed}",
            config.Id, status, smsSent, smsFailed);
    }

    public static string BuildMessage(string template, ShiftVolunteerDto shift, DateTime targetDate)
    {
        var firstName = shift.FirstName ?? "";
        var fullName = !string.IsNullOrEmpty(shift.FirstName) && !string.IsNullOrEmpty(shift.LastName)
            ? $"{shift.FirstName} {shift.LastName}"
            : shift.MappingName;
        var dateStr = targetDate.ToString("dd/MM/yyyy");
        var dayName = GetHebrewDayName(targetDate.DayOfWeek);

        return template
            .Replace("{שם}", firstName)
            .Replace("{שם מלא}", fullName)
            .Replace("{תאריך}", dateStr)
            .Replace("{יום}", dayName)
            .Replace("{משמרת}", shift.ShiftName)
            .Replace("{רכב}", shift.CarId);
    }

    public static string BuildLocationText(ShiftVolunteerDto shift)
    {
        if (string.IsNullOrEmpty(shift.LocationName))
            return "";

        var text = !string.IsNullOrEmpty(shift.LocationCity)
            ? $"\nהניידת נמצאת ב{shift.LocationCity} ({shift.LocationName})"
            : $"\nהניידת נמצאת אצל {shift.LocationName}";

        if (!string.IsNullOrEmpty(shift.LocationNavigation))
            text += $" ,נווט {shift.LocationNavigation}";

        return text;
    }

    public async Task<object> SendLocationUpdateAsync(DateTime date, string shiftName, string carId)
    {
        var dateStart = date.Date.ToString("o");
        var dateEnd = date.Date.AddDays(1).ToString("o");

        var eligibleShifts = await _db.Db.FetchAsync<ShiftVolunteerDto>(
            @"SELECT s.Id AS ShiftId, s.ShiftDate, s.ShiftName, s.CarId,
                     v.Id AS VolunteerId, v.FirstName, v.LastName, v.MappingName, v.MobilePhone,
                     s.LocationId,
                     COALESCE(l.Name, s.CustomLocationName) AS LocationName,
                     COALESCE(l.Navigation, s.CustomLocationNavigation) AS LocationNavigation,
                     l.City AS LocationCity
              FROM Shifts s
              JOIN Volunteers v ON s.VolunteerId = v.Id
              LEFT JOIN Locations l ON s.LocationId = l.Id
              WHERE s.ShiftDate >= @0 AND s.ShiftDate < @1
                AND s.ShiftName = @2 AND s.CarId = @3
                AND v.ApproveToReceiveSms = 1
                AND v.MobilePhone IS NOT NULL
                AND v.MobilePhone != ''",
            dateStart, dateEnd, shiftName, carId);

        var smsSent = 0;
        var smsFailed = 0;

        foreach (var shift in eligibleShifts)
        {
            try
            {
                var locationText = BuildLocationText(shift).TrimStart('\n');
                if (string.IsNullOrEmpty(locationText)) continue;

                var message = $"עדכון מיקום הניידת:\n{locationText}\nמשמרת נעימה";
                var result = await _smsProvider.SendSmsAsync(shift.MobilePhone!, message);

                await _db.SmsLog.InsertAsync(new SmsLog
                {
                    ShiftId = shift.ShiftId,
                    SentAt = DateTime.UtcNow,
                    Status = result.Success ? MagavConstants.SmsStatuses.Success : MagavConstants.SmsStatuses.Fail,
                    Error = result.Error,
                    ReminderType = MagavConstants.ReminderTypes.LocationUpdate
                });

                if (result.Success) smsSent++;
                else smsFailed++;
            }
            catch (Exception ex)
            {
                smsFailed++;
                _logger.LogError(ex, "Error sending location update SMS for ShiftId={ShiftId}", shift.ShiftId);
            }
        }

        return new { SmsSent = smsSent, SmsFailed = smsFailed };
    }

    public static string GetHebrewDayName(DayOfWeek day) => day switch
    {
        DayOfWeek.Sunday => "יום א׳",
        DayOfWeek.Monday => "יום ב׳",
        DayOfWeek.Tuesday => "יום ג׳",
        DayOfWeek.Wednesday => "יום ד׳",
        DayOfWeek.Thursday => "יום ה׳",
        DayOfWeek.Friday => "יום ו׳",
        DayOfWeek.Saturday => "שבת",
        _ => "לא ידוע"
    };
}

/// <summary>
/// DTO for the shift+volunteer join query used by the scheduler.
/// </summary>
public class ShiftVolunteerDto
{
    public int ShiftId { get; set; }
    public DateTime ShiftDate { get; set; }
    public string ShiftName { get; set; } = string.Empty;
    public string CarId { get; set; } = string.Empty;
    public int VolunteerId { get; set; }
    public string? FirstName { get; set; }
    public string? LastName { get; set; }
    public string MappingName { get; set; } = string.Empty;
    public string? MobilePhone { get; set; }
    public int? LocationId { get; set; }
    public string? LocationName { get; set; }
    public string? LocationNavigation { get; set; }
    public string? LocationCity { get; set; }
}
